compiler/src/org.graalvm.compiler.java/src/org/graalvm/compiler/java/BytecodeParser.java

```diff
@@ -1237,11 +1237,30 @@ protected void handleUnresolvedCheckCast(JavaType type, ValueNode object) {
      */
     protected void handleUnresolvedInstanceOf(JavaType type, ValueNode object) {
         assert !graphBuilderConfig.unresolvedIsError();
-        AbstractBeginNode successor = graph.add(new BeginNode());
```

Comment:
Was this change necessary? Or was it just for symmetry with the SVM code?

This structure will just end up being turned into a `FixedGuardNode`, i think should simplify this to just emitting a `FixedGuardNode` right away (like the checkcast case). Unless it fixes some issue i'm missing.

Reply:
> Was this change necessary? Or was it just for symmetry with the SVM code?

Only for symmetry.

> This structure will just end up being turned into a FixedGuardNode, i think should simplify this to just emitting a FixedGuardNode right away (like the checkcast case). Unless it fixes some issue i'm missing.

I'll try that instead.

Reply:
`FixedGuardNode` is undocumented, can you explain how I am expected to use it?

Reply:
`FixedGuardNode` is a control-flow node that ensures its condition input is true and deoptimizes if it's false.
You don't need to specify a probability since the condition is expected to be true

Reply:
>  You don't need to specify a probability since the condition is expected to be true.

My expectation in this situation is that most references input to `instanceof` will be non-null, and that is the opposite of what `FixedGuardNode` assumes when `IsNullNode.create(object)` is provided as a condition.

Reply:
I general that's true but the fact that the type is not resolved indicates that either the type cannot be resolved or it was never necessary to resolve it (we always get null). We don't really support the first case beyond correctness (i.e., we always deopt), and in the second case the implicit probability of the fixed guard node is correct.

If you have to give anything more than a very low probability to a deopt, something is wrong. That can happen for some cases that we don't really support in compiled mode when the compiler is used as a JIT.

Reply:
I followed your suggestion to mirror the implementation of `checkcast`, but I kept comments.